File: src/main/java/net/nodus/security/auth/OAuth2LoginSuccessHandler.java
package net.nodus.security.auth;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import net.nodus.database.account.OAuthProvider;
import net.nodus.database.account.UserAccount;
import net.nodus.database.account.UserAccountRepository;
import net.nodus.global.common.exception.GlobalException;
import net.nodus.security.auth.dto.OAuthResponse;
import net.nodus.security.auth.service.JwtTokenService;
import net.nodus.security.auth.service.RefreshTokenService;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.core.oidc.user.OidcUser;
import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class OAuth2LoginSuccessHandler implements AuthenticationSuccessHandler {

    private final UserAccountRepository userAccountRepository;
    private final JwtTokenService jwtTokenService;
    private final RefreshTokenService refreshTokenService;
    private final ObjectMapper objectMapper;

    @Override
    public void onAuthenticationSuccess(
        HttpServletRequest request,
        HttpServletResponse response,
        Authentication authentication
    ) throws IOException {
        OidcUser oidcUser = (OidcUser) authentication.getPrincipal();

        UserAccount user = findOrCreateUser(oidcUser);
        String accessToken = jwtTokenService.createAccessToken(user);
        RefreshTokenService.IssuedRefreshToken refreshToken = refreshTokenService.issue(user);

        response.setContentType("application/json");
        response.setCharacterEncoding("UTF-8");

        OAuthResponse.OAuthLoginResponse oAuthLoginResponse = OAuthResponse.OAuthLoginResponse.builder()
            .accessToken(accessToken)
            .refreshToken(refreshToken.token())
            .build();

        objectMapper.writeValue(
            response.getWriter(),
            oAuthLoginResponse
        );
    }

    private UserAccount findOrCreateUser(OidcUser oidcUser) {
        String providerId = oidcUser.getSubject();
        String email = oidcUser.getEmail();
        if (email == null) {
            throw new GlobalException.ExternalApiFailed("Google account email is missing");
        }
        String name = oidcUser.getFullName() == null ? email : oidcUser.getFullName();

        return userAccountRepository
            .findByProviderAndProviderIdAndDeletedAtIsNull(OAuthProvider.GOOGLE, providerId)
            .map(existingUser -> {
                existingUser.setEmail(email);
                existingUser.setName(name);
                return userAccountRepository.save(existingUser);
            })
            .orElseGet(() -> userAccountRepository.save(
                new UserAccount(email, name, OAuthProvider.GOOGLE, providerId)
            ));
    }
}

File: src/main/kotlin/net/nodus/auth/OAuth2LoginSuccessHandler.kt
package net.nodus.auth

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import net.nodus.account.OAuthProvider
import net.nodus.account.UserAccount
import net.nodus.account.UserAccountRepository
import net.nodus.auth.service.JwtTokenService
import net.nodus.auth.service.RefreshTokenService
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.core.oidc.user.OidcUser
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.stereotype.Component
import com.fasterxml.jackson.databind.ObjectMapper
import net.nodus.common.exception.GlobalException

@Component
class OAuth2LoginSuccessHandler(
    private val userAccountRepository: UserAccountRepository,
    private val jwtTokenService: JwtTokenService,
    private val refreshTokenService: RefreshTokenService,
    private val objectMapper: ObjectMapper,
) : AuthenticationSuccessHandler {

    override fun onAuthenticationSuccess(
        request: HttpServletRequest,
        response: HttpServletResponse,
        authentication: Authentication
    ) {
        val oidcUser = authentication.principal as OidcUser

        val user = findOrCreateUser(oidcUser)
        val accessToken = jwtTokenService.createAccessToken(user)
        val refreshToken = refreshTokenService.issue(user)

        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"

        objectMapper.writeValue(
            response.writer,
            OAuthLoginResponse(
                accessToken = accessToken,
                refreshToken = refreshToken.token,
            )
        )
    }

    private fun findOrCreateUser(oidcUser: OidcUser): UserAccount {
        val providerId = oidcUser.subject
        val email = oidcUser.email
            ?: throw GlobalException.ExternalApiFailed("Google account email is missing")
        val name = oidcUser.fullName ?: email

        val existingUser = userAccountRepository.findByProviderAndProviderIdAndDeletedAtIsNull(
            provider = OAuthProvider.GOOGLE,
            providerId = providerId
        )

        if (existingUser != null) {
            existingUser.email = email
            existingUser.name = name
            return userAccountRepository.save(existingUser)
        }

        return userAccountRepository.save(
            UserAccount(
                email = email,
                name = name,
                provider = OAuthProvider.GOOGLE,
                providerId = providerId,
            )
        )
    }
}

data class OAuthLoginResponse(
    val accessToken: String,
    val refreshToken: String,
    val tokenType: String = "Bearer",
)
